content/blog/optimizing-retrieval-with-hyde/index.md

```diff
@@ -0,0 +1,265 @@
+featured_image: thumbnail.png
+featured_image_caption: 
+images: ["blog/optimizing-retrieval-with-hyde/thumbnail.png"]
+toc: True
+date: 2024-02-28
+last_updated: 2024-02-28
+authors:
+  - David Batista
+tags: ["Generative AI", "Retrieval"]
+cookbook: using_hyde_for_improved_retrieval.ipynb
+---
+
+[Hypothetical Document Embeddings (HyDE)](https://docs.haystack.deepset.ai/v2.0/docs/hypothetical-document-embeddings-hyde) is a technique that allows us to generate “fake” hypothetical documents given a query and originates from the paper “[Precise Zero-Shot Dense Retrieval without Relevance Labels](https://aclanthology.org/2023.acl-long.99/)”. Based on the findings in the paper, this is done 5 times, and then we encode each hypothetical document into an embedding vector and average them. 
+
+In this article, we will see how we can use this approach to improve retrieval, and how we can easily incorporate it into Haystack by creating a [custom component](https://docs.haystack.deepset.ai/v2.0/docs/custom-components) that implements HyDE. 
+
+> To learn more about how HyDE works, and where it's useful, check out our guide on [Hypothetical Document Embeddings (HyDE)](https://docs.haystack.deepset.ai/v2.0/docs/hypothetical-document-embeddings-hyde) 
+
+## Build a Pipeline to Create Hypothetical Document Embeddings
+
+First, let's build a simple pipeline to generate these hypothetical documents. To do so, we will use the following Haystack components:
+
+-   [`PromptBuilder`](https://docs.haystack.deepset.ai/v2.0/docs/promptbuilder) and [`OpenAIGenerator`](https://docs.haystack.deepset.ai/v2.0/docs/openaigenerator) to query an instruction-following language model and generate hypothetical documents.
+-   [`SentenceTransformersDocumentEmbedder`](https://docs.haystack.deepset.ai/v2.0/docs/sentencetransformersdocumentembedder) encodes the hypothetical documents into vector embeddings.
+-  [`OutputAdapter`](https://docs.haystack.deepset.ai/v2.0/docs/outputadapter) to adapt the output of the `Generator` to be compatible with the input of the `SentenceTransformersDocumentEmbedder`, which expects `List[Document]`
+
+> To use the `OpenAIGenerator`, you need to set your `OPENAI_API_KEY`
+> ```python
+> export OPENAI_API_KEY="secret_string"
+>```
+
+We first build a way to query an instruction-following language model to generate hypothetical documents.
+
+```python
+from haystack.components.generators.openai import OpenAIGenerator
+from haystack.components.builders import PromptBuilder
+
+generator = OpenAIGenerator(model="gpt-3.5-turbo",
+							generation_kwargs={"n": 5, "temperature": 0.75, "max_tokens": 400},
+)
+
+template="""Given a question, generate a paragraph of text that answers the question.
+			Question: {{question}}
+			Paragraph:"""
+prompt_builder = PromptBuilder(template=template)
+```
+
+This will output a list of hypothetical documents. We then use the `SentenceTransformersDocumentEmbedder` to encode these hypothetical documents into embeddings.
```

Comment:
I would mention the 5 docs here instead:

`This will output a list of 5 hypothetical documents,  the same number the authors used for the experiments in the paper. We then use the `SentenceTransformersDocumentEmbedder` to encode these hypothetical documents into embeddings.`